### ISSUE TITLE
0.100.x: AdjustToContents slow

### ISSUE THREAD

**Issue by eithe** [ClosedXML/ClosedXML#1991]
Hi, and thanks for taking the time to address the memory consumption with the 0.100.x releases!

We run benchmarks (using BenchmarkDotNet) on our Excel functionality, and we do indeed see a much smaller memory footprint with the 0.100.x release :)

**Do you want to request a *feature* or report a *bug*?**
- [x] Bug
- [ ] Feature
- [ ] Question

**Did you test against the latest CI build?**
- [x] Yes
- [ ] No

**Version of ClosedXML**

0.100.3

**What is the current behavior?**

We run the following command after our data inserts, and this line takes about 50s on the 0.100.x release vs about 2s on the 0.96 release. This is for an Excel sheet of 10.000 rows and ~70 cols, running on Windows.

`worksheet.Columns().AdjustToContents(minWidth: 0, maxWidth: 150);`

Do you also see the same? Perhaps we are doing something that we shouldn't?

**What is the expected behavior or new feature?**

That it would take a few seconds (max) like before.

**Is this a regression from the previous version?**

Yes, at least from the 0.96 release

## Reproducibility

**Code to reproduce problem:**

Make any sheet with around 10k rows and 50-100 cols, then try to run
```c#
worksheet.Columns().AdjustToContents(minWidth: 0, maxWidth: 150);
```

- [ ] I attached a sample spreadsheet.  (You can drag files on to this issue)

## Workaround
As a workaround, we have opted to use startRow and endRow, that completes in about ~1s-2s.

`worksheet.Columns().AdjustToContents(startRow: 1, endRow: Math.Min(rowCount, 100), minWidth: 0, maxWidth: 150);`

**Comment by Karql** [comment]
Hi,

I wanted to add the same issue but you were first 😉

It's looks like problem occurs after this change: https://github.com/ClosedXML/ClosedXML/commit/e860f86dfd6ecdd051aef766c0c0427c6c1690cb#diff-0f4789a22b967096e05186b2cb42cde2107719e3523637003372cb9b4e462c37R91

Sample to reproduce it:

```c#
var sentences = new[]
{
    "Lorem ipsum dolor sit amet, consectetur adipiscing elit.",
    "Donec interdum consectetur leo, vel pretium sapien bibendum id.",
    "Ut tristique, dolor in feugiat tristique, diam nisl varius risus, et luctus risus purus eget elit.",
    "Pellentesque nec tincidunt sapien.",
    "In semper pulvinar sapien, vel dignissim massa molestie sed.",
    "Donec efficitur iaculis est in posuere.",
    "Phasellus quis mattis est.",
    "Integer facilisis sem sem, eu sagittis nibh eleifend ac.",
    "Vivamus ultricies tortor condimentum consectetur ornare.",
    "Aliquam sollicitudin rutrum erat, a placerat tortor vestibulum eget.",
    "In at ante at lorem tristique facilisis sed in neque.",
    "Sed pretium ipsum nulla, vel iaculis ex interdum at.",
    "In scelerisque, orci ac sollicitudin aliquam, enim diam convallis nibh, vitae auctor tellus dolor in leo.",
    "Nulla ultricies arcu arcu.",
    "Nulla tortor metus, viverra eu consectetur eu, vulputate vel urna.",
    "Etiam lacinia sapien ornare suscipit tincidunt.",
    "Vivamus mollis nunc in ante porta aliquam.",
    "Vivamus sed condimentum elit, at hendrerit mi.",
    "Vivamus porttitor tempus fermentum.",
    "In pharetra mauris non posuere interdum.",
    "Duis ullamcorper in ante et dapibus.",
    "Donec lacinia augue a ultrices luctus.",
    "Duis non rutrum eros.",
    "Aliquam pharetra euismod convallis.",
    "Cras tristique turpis vel odio maximus, sit amet vulputate nisl vehicula.",
    "Maecenas neque urna, rutrum accumsan tempor ut, consectetur volutpat diam.",
    "Donec pretium, mi sed vulputate tristique, nibh est dictum magna, ut efficitur metus ante ac quam.",
    "Mauris tincidunt tortor eget hendrerit porta.",
    "Lorem ipsum dolor sit amet, consectetur adipiscing elit.",
    "Etiam interdum finibus turpis.",
    "Aenean viverra ut est quis porttitor.",
    "Phasellus vitae porttitor dolor.",
    "Sed ornare ligula ut euismod vulputate.",
    "Integer ac imperdiet turpis.",
    "Quisque ante risus, pharetra sit amet turpis non, tempor bibendum enim.",
    "In sed nisl vestibulum, cursus purus sit amet, scelerisque velit.",
    "Maecenas ut eleifend orci.",
    "In facilisis quam quis dolor porta, eget blandit turpis consequat.",
    "Maecenas vel nisi in urna blandit faucibus at ut elit.",
    "Cras ut turpis lacus.",
    "Integer at urna quis leo sollicitudin pellentesque.",
    "Nam quis blandit leo.",
    "Nulla a ante hendrerit purus viverra auctor.",
    "Aenean arcu lorem, pulvinar sed elit ut, euismod placerat orci.",
    "Maecenas sed odio neque.",
    "Curabitur tincidunt, urna eget venenatis pulvinar, massa nisi dictum magna, nec ultricies tellus est sed enim.",
    "In pulvinar libero ut nisl ultrices, ac cursus mauris volutpat.",
    "Vestibulum interdum eleifend lorem eu hendrerit.",
    "Praesent quis posuere libero, id tincidunt lorem.",
    "Mauris in congue tortor.",
    "Sed sed aliquam leo.",
    "Donec imperdiet felis ac elit tincidunt suscipit.",
    "Morbi ultricies justo sed rutrum mattis.",
    "Sed auctor odio in augue placerat, non egestas enim pellentesque.",
    "Proin dignissim mollis mauris nec fringilla.",
    "Vivamus turpis leo, porttitor vitae lectus in, euismod suscipit mauris.",
    "Donec tempus sapien sed libero lobortis, commodo interdum velit venenatis.",
    "Proin nisl eros, consectetur a efficitur et, iaculis dignissim nulla.",
    "Cras tempor tortor sit amet egestas mattis.",
    "Integer tempus consequat ipsum, sagittis tristique felis aliquet id.",
    "Nunc eget felis sapien.",
    "In id ex vitae elit viverra finibus id eget dolor.",
    "Proin ullamcorper neque quis dictum gravida.",
    "Morbi a efficitur tellus.",
    "Integer lobortis varius massa.",
    "Phasellus malesuada accumsan felis, in ultricies elit consectetur a.",
    "Nunc aliquam nibh sit amet turpis ullamcorper efficitur.",
    "Donec rhoncus orci nulla, vel laoreet arcu sagittis in.",
    "Vestibulum sollicitudin venenatis sodales.",
    "Pellentesque gravida varius sapien ac pulvinar.",
    "Quisque sit amet blandit justo.",
    "Nullam fringilla nisi sed neque fermentum laoreet.",
    "Maecenas sed ullamcorper felis."
};

var testFunc = () =>
{
    var wb = new XLWorkbook();
    var sheet = wb.AddWorksheet("Test");

    var rowsCount = 1000;
    var sentencesCount = sentences.Length;


    for (int row = 1; row <= rowsCount; row++)
    {
        sheet.Cell(row, 1).SetValue(sentences[row % sentencesCount]);
    }

    var stopWatch = new Stopwatch();

    stopWatch.Start();
    sheet.Columns().AdjustToContents();
    stopWatch.Stop();

    return stopWatch.ElapsedMilliseconds;
};

var testResults = Enumerable.Range(1, 10).Select(x => new { test = x, time = testFunc() }).ToArray();
```

Result for 0.96.0:
![image](https://user-images.githubusercontent.com/173098/215396881-f4906542-74ca-4a6f-87df-9958f40cf462.png)

Result for 0.97.0:
![image](https://user-images.githubusercontent.com/173098/215396922-a0a22451-4b11-494c-8001-c82d9c0c2bc8.png)

Diagnostics for `sheet.Columns().AdjustToContents();` in 0.97.0:

![image](https://user-images.githubusercontent.com/173098/215397749-d532fe74-c34b-41d1-a04f-2d79fa00b180.png)

Best regards,
Mateusz

**Comment by jahav** [comment]
The problem here is 
* Unlike deprecated System.Drawing.Common, SixLabors.Fonts has to load fonts for each process. System.Drawing was calling native GDI/GDI+ through p/Invoke and all fonts were already in memory, ready to be used. That is the startup cost in first test.
* _SixLabors.Fonts_ `TextMeasurer` uses proper font shaping engine. Shaping engine is the thing that correctly deduces where each glyph should be to determine final position. 

I don't think I can do anything with the startup cost. That is up to SixLabors.Fonts. E.g. Calibri is 1.6 MB font and it takes up time to load it and parse. From what I have seen, SixLabors.Fonts basically just reads tables with minimal processing. Font files are mostly a bunch of lookup tables anyway, so there isn't some cpu-intensive processing. I had some improvements by reading it to memory, but it was flaky (unreproducible, only sometimes) + it didn't make sense since files are already cached in memory by memory blocks.

But I can do something with the shaping engine, because Excel doesn't actually use it (at least as far as I can see). No glyph substitution, no ligarutes, no bidirectional text(e.g. one part is from left-to-right script and other with right-to-left), pretty sure GPOS is not used, no kerning.... It seems that Excel basically takes a unicode codepoint and gets advance width of a glyph (even anceint Win32 GDI at least takes ligatures into account, not Excel). That is good, because it is really fast. It is also not accurate, but nobody probably complained (or it was dumped due to high computation requirement which ClosedXMl is seeing)

> Computer joke from the 80's: 6502: Hey guys! What's 2+2? Intel: 5! Motorola: Hmmm... 4? Intel: Yes, but I was faster!

![image](https://user-images.githubusercontent.com/7634052/218313229-ad428f0c-9a10-44b7-bb51-44012dd02f14.png)

Basically say that advance width (converted to pixels, round px) is the width of a codepoint SixLabors has `font.FontMetrics.GetGlyphMetrics` method to get glyph metrics, include advance width in font units. 

Adjust to content feature in Excel is fun (by fun, I mean ungodly undocumented mess with inconsistent terminology). Over past 6 months, I went so deep into fonts and rendering that I could probably start my own font library (not that I would). Output is mostly at https://github.com/ClosedXML/ClosedXML/wiki/Cell-Dimensions

**Comment by Karql** [comment]
@jahav have you tried [IronSoftware.System.Drawing](https://github.com/iron-software/IronSoftware.System.Drawing) ?
Some time ago during [ocr](https://ironsoftware.com/csharp/ocr/) I came across info that IronSoftware release their version of `System.Drawing`.
Maybe that would be the solution.

**Comment by jahav** [comment]
That is only an adapter package between System.Drawing, SixLabors.Fonts/ImageSharp and SkiaSharp. Useful for migration away from System drawing, but doesn't contain the features.

**Comment by JimBobSquarePants** [comment]
If you are not already, you should cache loaded fonts, so they are not reloaded for each operation. Measurement itself should be fast but loading a font is intensive and performance varies wildly depending on the size of the font.

**Comment by jahav** [comment]
@JimBobSquarePants Thanks for the hint. I am already caching the Font object. I think that the best solution for performance in this case (and likely many others) is to use multiple threads - something that will require some deep thought.

**Comment by eithe** [comment]
Imho it's more than fast enough now, and generally we are really satisfied with the latest development on this project. Keep it up, @jahav 👏 You should setup GitHub Sponsor when you get the time :)
